io/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -471,22 +455,16 @@ private object RTS {
                       }
                     } else {
                       // Error caught:
-                      // FIXME this cast is not safe, `err` is defined as `E => IO[E2, B]`
-                      val handledIo = errorHandler(error).asInstanceOf[IO[E, Any]]
-
-                      if (finalizer eq null) {
-                        // No finalizer to run:
-                        curIo = handledIo
+                      curIo = nextInstr[E](error, stack)
 
-                      } else {
+                      if (finalizer ne null) {
                         // Must run finalizer first:
                         val finalization = dispatchErrors(finalizer)
-                        val completer    = handledIo
 
                         // Do not interrupt finalization:
                         this.noInterrupt += 1
 
-                        curIo = ensuringUninterruptibleExit(finalization[E] *> completer)
+                        curIo = ensuringUninterruptibleExit(finalization[E] *> curIo)
```

Comment:
I think this should actually be:

```scala
curIo = ensuringUninterruptibleExit(finalization[E]) *> curIo
```

Only the finalization needs to be protected from interruption, not the error handler itself.

Reply:
Hmmm, I've tried with:

```
                        curIo = ensuringUninterruptibleExit(finalization[E]) *> curIo
```

And got same error. This used to be the same when the completer was the hardcoded `now` of `Left`